README.rst
=======
QoS_NSF
=======


* Overview:
The performance of a wireless smart sensor network (WSSN) is quantified by user-specific "Quality of Service (QoS)" metrics. 
A practical WSN deployed in the field is almost always built with redundancy. It is achieved either through hardware or algorithms or both. 
Therefore, the minimal QoS can be maintained through different QoS node configuration.

* The main objective of this ongoing project is to develop a framework for ensuring a desired QoS based on the system identification quality from a real-life sensor network.
Initially, the accuracy of the estimated mode shape is considered as a quality index. Modal analysis is required to define the quality of mode shape
obtained from different sensor node configuration. Then metrices can be introduced to ensure minimal quality throughout the network operation.

* Modal Analysis:
For real-life sensor node configurations, a simply supported bridge is considered. Output-based system identification is considered for modal analysis.
There are different output-based modal analysis methods available e.g. Covariance driven stochastic system identification, Data driven system identification method etc.
For this analysis only Covariance driven stochastic system identification i.e. obtaining mode shapes from ambient vibration due to traffic load is considered.



Features
--------

* TODO

Credits
-------

1. This project package was created from the template of Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_.

    .. _Cookiecutter: https://github.com/audreyr/cookiecutter
    .. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage

2. To create python package for modal analysis of beam, the algorithm of strid_ (basically developed for shear frame) is followed:
     
     .. _strid: https://github.com/Gunnstein/strid

HISTORY.rst
=======
History
=======

0.1.0 (2022-10-02)
------------------

* First release on PyPI.


qos-pypackage/csi-beam/sysid/test.py
# Import the response data from different node configuration
data = np.load("node9data.npz")
y = data["y"]
fs = data["fs"]
true_f = data["true_frequencies"]
true_xi = data["true_damping"]
true_modeshapes = data["true_modeshapes"]


ssid = sysid.CovarianceDrivenStochasticSID(y, fs)

# Testing Covariance driven stochastic system ID



qos-pypackage/csi-beam/data-gen.py
"""Console script for qos."""
#%matplotlib nbagg

import numpy as np
import matplotlib.pyplot as plt
import scipy.signal
import sysid
from scipy import io

''' ----- Create a sensor network for a beam with 10 sensors spaced symmetrically from each other---
input for simulating beam geometry
n=# of sensor nodes activated in the network excluding the two support nodes
Lt= Total length of the beam (feet)
Im= Moment of inertia (in^4)
E= modulus of elasticity (psi)
m_bar= mass density per unit length (lb.sec^2/in/in)'''
#pos=np.array([0,10,12.5,16.7,25,30,37.5,50,62.5,70,75,83.3,87.5,90,100])*12
pos=np.array([0,5,10,15,20,25,30,35,40,45,50,55,60,65,70,75,80,85,90,95,100])*12
s=np.array([0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,1,1])
pos_n=pos*s
pos_n=pos_n[np.nonzero(pos_n)]
pos_n=np.r_[0,pos_n]
n=len(pos_n)-2

data = np.load("node19data.npz")
y = data["y"]
fs = data["fs"]
s=s[1:-1]
y_n=y*np.array([s]).T
y_n=y_n[~np.all(y_n == 0, axis=1)]

bm = sysid.utils.Beam(pos_n, Im=100, E=6.58*1e6, m_bar=0.1)
"""
n=len(pos)-2

bm = sysid.utils.Beam(pos, Im=100, E=6.58*1e6, m_bar=0.1)


# Determine the time discretization and period

Tmax = 1. / sysid.utils.w2f(bm.get_natural_frequency(1))
fmax = sysid.utils.w2f(bm.get_natural_frequency(n))

'''
T = 1000 * Tmax
fs = 5 * fmax'''

T=180.0
fs=512.0
t = np.arange(0., T, 1 / fs)

# Define loads on system
## Unmeasureable: Stochastic loads on all nodes (Process noise)
w = np.random.normal(size=(n, t.size))*1e1

## Load matrix, f
F = w.copy() # checked random noise looks ok

# Simulate response, displacement at each node measured
_, _, y0 = bm.simulate(t, F)

noise_std = y0.std()

# Add measurement noise
v = np.random.normal(size=y0.shape)* noise_std*15*1e-2
y = y0 + v

plt.figure("Displacement measured last node")
plt.plot(t, y[-1], label="w/noise")
plt.plot(t, y0[-1], label="wo/noise")
plt.legend()
plt.xlabel("Time (s)")
plt.ylabel("Amplitude")

plt.figure("PSD of Displacement at last node")
for yi in [y[-1], y0[-1]]:
    freqs, Pyy = scipy.signal.csd(yi,yi, fs, nperseg=2 ** 12)
    plt.semilogy(freqs, Pyy)
    plt.xlim([0,fmax+10])

for m in range(n):
    plt.axvline(sysid.utils.w2f(bm.get_natural_frequency(m+1)), alpha=.2)

plt.ylabel('PSD')
plt.xlabel('Frequency (Hz)')
plt.show()
"""
bm.set_rayleigh_damping_matrix([bm.get_natural_frequency(1), bm.get_natural_frequency(n)], [.05]*2)

true_frequencies = np.array([bm.get_natural_frequency(i)/(2*np.pi) for i in range(1, n+1)])
true_damping = np.array([bm.get_rayleigh_damping_ratio(i) for i in range(1, n+1)])
true_modeshapes = np.array([bm.get_mode_shape(i) for i in range(1, n+1)])

#io.savemat("nodepeakdata.mat",{'y':y,'fs':fs,'f':true_frequencies,'xi':true_damping,'phi':true_modeshapes,'posSensors':pos})

#print(np.shape(y))

np.savez('node19-11-1data',
         y=y_n, fs=fs,
         true_frequencies=true_frequencies,
         true_damping=true_damping,
         true_modeshapes=true_modeshapes,
         sensor_pos=pos_n
         )
print(np.size(true_frequencies))
print(np.shape(true_modeshapes))
print(np.shape(y_n))
print(np.shape(y))


qos-pypackage/csi-beam/sysid/utils.py
"""Utilities"""
import numpy as np
from scipy import linalg,signal
import scipy

__all__ = ["find_rayleigh_damping_coeffs", "get_frequency_vector",
           "modal_assurance_criterion", "w2f", "f2w", "norm2",
           "modal_scale_factor", "modal_phase_collinearity",
           "mean_phase", "mean_phase_deviation", "find_psd_matrix",
           "find_positive_psd_matrix", "find_frf_matrix",
           "accelerance2receptance", "receptance2accelerance", "Mode"]


def find_rayleigh_damping_coeffs(w, damping_ratios):
    """Rayleigh damping coefficients from target freqs and damping ratio

    Rayleigh damping is defined by the following relation

    .. math::

        C = aM + bK

    where :math:`C` is the damping matrix, :math:`M` is the mass
    matrix and :math:`K` is the stiffness matrix. The damping ratio
    :math:`\\xi` at angular frequency :math:`\\omega` is then given by
    the following equation:

    .. math::

       \\xi = \\frac{1}{2}(\\frac{a}{\\omega} + b\\omega)

    The damping coefficients is determined by specifying the desired
    damping ratio at two or more frequencies by the least square
    method.

    Arguments
    ---------
    w : 1darray
        Angular frequencies (rad/s) where the damping ratios are
        specified
    damping_ratios : 1darray
        The damping ratios (c / c_cr) at the specified frequencies.

    Returns
    -------
    a, b : float
        Rayleigh dampping coefficients
    """
    A = .5 * np.array([[1 / wn, wn]
                       for wn in w])
    return np.linalg.lstsq(A, damping_ratios, rcond=None)[0]


def w2f(w):
    "Convert angular frequency (rad/s) to frequency (Hz)"
    return w / (2*np.pi)


def f2w(f):
    "Convert frequency (Hz) to angular frequency (rad/s)"
    return f * (2*np.pi)


def modal_scale_factor(u, v):
    """Determine the scale factor between u and v

    Find the scale factor :math:`\\alpha` between :math:`u` and
    :math:`v` such that :math:`\\alpha u` has similar length and
    direction as :math:`v`.

    Argument
    --------
    u, v : 1darray
        Vectors to find the modal scale factor between

    Return
    ------
    complex
        The modal scale factor that makes the two vectors of
        similar length and phase.

    """
    u, v = np.asarray(u), np.asarray(v)
    return u.conj().dot(v) / np.linalg.norm(u)**2


def modal_phase_collinearity(u):
    """Modal phase collinearity of mode vector u

    Mode phase collinearity (MPC) quantifies the complexity of a
    mode vector. For classical normal modes, all dofs in a
    structure vibrate in phase with one another such that a maximum
    modal value is reached at the same instant for all dofs. MPC
    ranges from 0 to 1, where MPC=1 indicates that all dofs are in
    phase, and zero indicates out of phase dofs.

    Arguments
    ---------
    u : 1darray[complex]
        Mode shape vector

    Returns
    -------
    float
        Modal phase collinearity of the mode vector

    References
    ----------
    [Pappa1992] Pappa RS, Elliott KB, Schenk A (1992)
        A consistent-mode indicator for the eigensystem realization
        algorithm. NASA Report TM-107607
    """
    u = np.asarray(u)
    S = np.cov(u.real, u.imag)
    l = np.linalg.eigvals(S)
    return (l[0]-l[1])**2/(l[0]+l[1])**2


def mean_phase(u):
    """Mean phase of mode vector u

    Mean phase (MP) is the angle of a linear line fitted through the
    mode shape in the complex plane.

    Arguments
    ---------
    u : 1darray[complex]
        Mode shape vector

    Returns
    -------
    float
        Mean phase of the mode vector

    References
    ----------
    [Reynders2012] Reynders, E., Houbrechts, J., De Roeck, G., 2012.
        Fully automated (operational) modal analysis. Mechanical
        Systems and Signal Processing 29, 228–250.
        https://doi.org/10.1016/j.ymssp.2012.01.007
    """
    u = np.asarray(u)
    U, s, VT = np.linalg.svd(np.c_[u.real, u.imag])
    V = VT.T
    return np.arctan2(V[0, 1], -V[1, 1])


def mean_phase_deviation(u):
    """Mean phase deviation of mode vector u

    Mean phase deviation (MPD) is the deviation in phase from the mean
    phase and is a measure of mode shape complexity.

    Arguments
    ---------
    u : 1darray[complex]
        Mode shape vector

    Returns
    -------
    float
        Mean phase deviation of the mode vector

    References
    ----------
    [Reynders2012] Reynders, E., Houbrechts, J., De Roeck, G., 2012.
        Fully automated (operational) modal analysis. Mechanical
        Systems and Signal Processing 29, 228–250.
        https://doi.org/10.1016/j.ymssp.2012.01.007
    """
    u = np.asarray(u)
    U, s, VT = np.linalg.svd(np.c_[u.real, u.imag])
    V = VT.T
    w = np.abs(u)
    num = u.real*V[1, 1] - u.imag*V[0, 1]
    den = np.sqrt(V[0, 1]**2+V[1, 1]**2)*np.abs(u)
    return np.sum(w*np.arccos(np.abs(num/den))) / np.sum(w)


def norm2(v):
    """Return the Euler norm of v.

    Return the Euler norm :math:`||v||_2`.

    Arguments
    ---------
    v : 1darray
        Vector to find the norm of.

    Returns
    -------
    float : Euler norm of v.
    """
    return np.linalg.norm(v, 2)


def get_frequency_vector(fs, n):
    #return np.linspace(0., fs/2, n)
    return np.linspace(0., fs/5, n)


def modal_assurance_criterion(u, v):
    """Calculate the modal assurance criterion (MAC)

    MAC is the square of the linear correlation between two mode shapes
    and is a measure of the similarity of two different modes which varies
    between 0 and 1. MAC=1 means a perfect correlation between two vectors
    exists, while MAC=0 means no linear correlation.

    Arguments
    ---------
    u, v : 1darray
        Mode shapes to check the MAC for.

    Returns
    -------
    float
        MAC between two vectors u and v.
    """
    u = np.asarray(u)
    v = np.asarray(v)
    return np.abs(u.conj().T@v)**2 / (u.conj().T@u*v.conj().T@v).real


def find_psd_matrix(x, y, **kwargs):
    """Calculate PSD matrix for x and y.

    Arguments
    ---------
    x, y : 2darray
       Measurement matrix where each row corresponds to the entire time
       series of a measurement channel.
    kwargs : dict, optional
       All keyword arguments are passed to the scipy.signal.csd,
       see docstring of scipy.signal.csd.

    Returns
    -------
    3darray
       PSD matrix where the first and second axis refers to the
       degree of freedom of `x` and `y`, respectively and the third axis
       refers to the frequency.

    See Also
    --------
    find_positive_psd_matrix :
        Positive lag psd matrix for frequency based operational modal
        analysis.
    """
    Pxy = np.array(
        [[scipy.signal.csd(xi, yj, **kwargs)[1]
          for yj in y] for xi in x])
    return Pxy


def find_positive_psd_matrix(x, y, nfft=2**9, window="rectangular"):
    """Calculate positive lag PSD matrix

    Calculate the positive lag PSD matrix as per [Cauberghe2004] (pg.
    47, fig. 3.8). Useful in combined deterministic-stochastic or
    stochastic frequency domain modal extraction methods such as LSCF
    and pLSCF.

    Arguments
    ---------
    x, y : 2darray
       Measurement matrix where each row corresponds to the entire time
       series of a measurement channel.
    nfft : int, optional
        Length of the block/segment used in estimation of PSD. Note that
        the number of samples from `x` and `y`, is half of `nfft`, i.e.
        `nperseg=nfft//2` and the segment is then zero padded.
    window : {'rectangular', 'exponential'}, optional
        Which window to apply to remove the negative lag coefficients.

    Returns
    -------
    3darray
       PSD matrix where the first and second axis refers to the
       degree of freedom of `x` and `y`, respectively and the third axis
       refers to the frequency.

    Raises
    ------
    ValueError
        Raised if `window` parameter is not `rectangular` or `exponential`.

    See Also
    --------
    find_psd_matrix :
        Conventional psd matrix.
    """
    Pxy = find_psd_matrix(x, y,
                          nperseg=nfft//2,
                          nfft=nfft,
                          noverlap=0,
                          window="boxcar",)
    Rxy = np.fft.irfft(Pxy)
    if window == "rectangular":
        win = scipy.signal.boxcar(Rxy.shape[2])
        win[Rxy.shape[2]//2:] *= 0.
    elif window == "exponential":
        tau = - Rxy.shape[2] / np.log(.01)
        win = scipy.signal.exponential(
            Rxy.shape[2], center=0, tau=tau, sym=False)
    else:
        raise ValueError(
            "`window` must be either `rectangular` or exponential`"
        )
    Rxy *= win
    return np.fft.rfft(Rxy)


def find_frf_matrix(u, y, estimator="H1", **kwargs):
    """Estimate the FRF matrix from input and output.

    Estimate the FRF matrix from input `u` and output `y`.

    Arguments
    ---------
    u, y : 2darray
        Input and output where each row contains the data points
        for a particular channel.

    estimator : str
        FRF estimator can be either `H1` or `H2`. The `H2`
        estimator requires an equal number of input and outputs
        to exist.

    kwargs :
       All keyword arguments are passed to the scipy.signal.csd,
       see docstring.

    Returns
    -------
    3darray
       FRF matrix where the first and second axis refers to the
       degree of freedom of `y` and `u`, respectively and the third axis
       refers to the frequency.

    Raises
    ------
    ValueError
        If the `H2` estimator is selected, but the number of inputs does not
        match the number of outputs.
    ValueError
        If estimator is not `H1` or `H2`.
    """
    if estimator.upper() == "H1":
        Sr = find_psd_matrix(u, u, **kwargs)
        Sl = find_psd_matrix(y, u, **kwargs)
    elif estimator.upper() == "H2":
        if u.shape[0] != y.shape[0]:
            raise ValueError(
                "# of inputs must match # of outputs for `H2` estimator."
            )
        Sr = find_psd_matrix(u, y, **kwargs)
        Sl = find_psd_matrix(y, y, **kwargs)
    else:
        raise ValueError(
            "FRF estimator must be either `H1` or `H2`."
        )
    H = np.moveaxis(np.array([
        np.linalg.solve(Sr[:, :, i].T, Sl[:, :, i].T).T
        for i in range(Sr.shape[2])]), 0, 2)
    return H


def accelerance2receptance(H, fs):
    """Convert accelerance to receptance.

    Convert the FRF matrix of accelerations, i.e. accelerance to a
    FRF matrix of displacements, i.e. receptance.

    .. math::

        H_u(\\omega) = \frac{H_\\ddot{u}(\\omega)}{\\omega^{2}}

    The operation is done in place and no value is returned.

    Arguments
    ---------
    H : 3darray
        Accelerance, first axis refers to the output,
        second axis to the input and third axis to the
        frequency.
    fs : float
        Sampling rate.
    """
    w = 2*np.pi*np.linspace(0., fs/2, H.shape[2])
    for i, wi in enumerate(w):
        if wi == 0.:
            continue
        H[:, :, i] /= -wi**2


def receptance2accelerance(H, fs):
    """Convert accelerance to receptance.

    Convert the FRF matrix of displacement, i.e. receptance to a FRF
    matrix of accelerations, i.e. accelerance.

    .. math::

        H_\\ddot{u}(\\omega) = H_u(\\omega) \\omega^{2}

    The operation is done in place and no value is returned.

    Arguments
    ---------
    H : 3darray
        Receptance, first axis refers to the output,
        second axis to the input and third axis to the
        frequency.
    fs : float
        Sampling rate.
    """
    w = 2*np.pi*np.linspace(0., fs/2, H.shape[2])
    for i, wi in enumerate(w):
        H[:, :, i] *= -wi**2


class Beam(object):
    def __init__(self, pos, Im, E, m_bar):
        """Create a simply supported beam object
        Define the beam with the mass `m` at each dof,
        stiffness `k`, and the number of sensors representing # of beam segments, see figure below.


                nd1         nd2       nd3.........                                              ndn-1    ndn
                 |   m, k   |   m, k  |   m, k  |   m, k  |   m, k  |   m, k  |   m, k  |   m, k  |   m, k  |
                **=========||========||========||========||========||========||========||========||========||**


        Arguments
        ---------
        n : int
            # of nodes/dofs for only vertical deflection (removing the 2 support nodes)
        K : float
            Global stiffness matrix
        M : float
            Global mass matrix
        """
        self.pos=pos
        self.n = len(pos)-2
        self.L = np.zeros((len(pos)-1),float)
        for ii in range(self.n + 1):
            self.L[ii] = pos[ii + 1] - pos[ii]
        self.Im = Im
        self.E = E
        self.m_bar =m_bar
        self.K=self._get_K()
        self.M=self._get_M()
        self.C = np.zeros_like(self.M)

    def _get_K(self):
        n= self.n
        E, I, L = self.E, self.Im, self.L
        K = np.zeros((n+2, n+2), float)

        for i in range(n+1):
            ke = np.array([[12 * E * I / L[i] ** 3, -12 * E * I / L[i] ** 3],
                           [-12 * E * I / L[i] ** 3, 12 * E * I / L[i] ** 3]])
            K[i:i + 2, i:i + 2] = K[i:i + 2, i:i + 2] + ke

        K=K[1:n+1,1:n+1]
        return K

    def _get_M(self):
        n= self.n
        m_bar, L = self.m_bar, self.L
        M = np.zeros((n+2, n+2), float)
        for i in range(n+1):
            me = m_bar * L[i] / 420 * np.array([[156, 54],
                                             [54, 156]])
            M[i:i + 2, i:i + 2] = M[i:i + 2, i:i + 2] + me

        M = M[1:n+1, 1:n+1]
        return M

    def get_natural_frequency(self, r):
        """Returns the analytical natural frequency of mode `r`
        Arguments
        ---------
        r : int
            Mode to return the frequency for.
        Returns
        -------
        float
            Natural frequency of mode `r` in rad/s
        """
        K=self.K
        M=self.M
        value,vector=linalg.eig(K,M)
        #omega=np.sort(np.sqrt(np.diag(value)))
        omega=np.sort((np.sqrt(value)).real)
        #omega = np.sort(omega)
        return omega[r-1]

    def get_mode_shape(self, r):
        """Returns the analytical mode shape of mode `r`
        Arguments
        ---------
        r : int
            Mode to return the mode shape for.
        Returns
        -------
        1darray
            Mode shape of mode `r`, the mode shape is normalized
            to have unit length.
        """
        value, vector = linalg.eig(self.K, self.M)
        value = np.array([(np.sqrt(value)).real])
        x = np.concatenate((value.T, vector.T), axis=1)
        x_sorted= x[x[:, 0].argsort()]
        a=x_sorted[:,1:]
        a=a.T
        return a[:,r-1]

    def set_rayleigh_damping_matrix(self, freqs, xis):
        """Set the damping matrix to the Rayleigh damping matrix
        Rayleigh damping is a classical damping matrix which defines
        the modal damping ratio x(w) by the following equation
            xi(w) = 1/2 * (a/w + b*w)
        where w is the frequency in rad/s and a and b are the Rayleigh
        damping coefficients. Specify the damping ratio for atleast two
        frequencies, if more than two damping ratios are specified, the
        Rayleigh damping coefficients are determined by least squre fitting.
        Arguments
        ---------
        freqs : 1darray
            Frequencies (rad/s) where the damping ratios are specified
        damping_ratios : 1darray
            The damping ratios (c / c_cr) at the specified frequencies.
        """
        a, b = find_rayleigh_damping_coeffs(freqs, xis)
        self._rayleigh_coeffs = (a, b)
        self.C = a*self.M + b*self.K

    def get_rayleigh_damping_ratio(self, r):
        """Returns the Rayleigh damping ratio of mode `r`
        It is assumed that Rayleigh damping is set, see
        set_rayleigh_damping_matrix method.
        Arguments
        ---------
        r : int
            Mode to return the damping ratio for.
        Returns
        -------
        float
            Damping ratio of mode `r`
        """

        a, b = self._rayleigh_coeffs
        w = self.get_natural_frequency(r)
        return .5*(a/w+b*w)

    @property
    def state_matrix(self):
        "CT State space  matrix (A)"
        C= self.C
        K, M=self.K,self.M
        Z = np.zeros_like(M)
        I = np.eye(M.shape[0])
        A11 = -np.linalg.solve(M, C)
        A12 = -np.linalg.solve(M, K)
        A = np.r_[np.c_[A11, A12],
                  np.c_[I, Z]]
        return A

    @property
    def input_influence_matrix(self):
        "CT State space input influence matrix (B)"
        K, M = self.K,self.M
        n=self.n
        return np.r_[np.linalg.solve(M, np.eye(n)),
                     np.zeros((n, n))]

    def get_state_space_matrices(self):
        '''Continous time state space matrices A, B, C, D
        C= output matrix of state-space system
        D= Feed through matrix of state-space system'''

        A = self.state_matrix
        B = self.input_influence_matrix
        n = self.n
        M=self.M
        O = np.zeros((n, n))
        I = np.eye(n)
        C = np.r_[A[:n, :],
                  np.c_[I, O],
                  np.c_[O, I]]
        D = np.r_[np.linalg.solve(M, I),
                  O,
                  O]
        return A, B, C, D

    def simulate(self, t, F=None, d0=None, v0=None):
        """Obtain system response to load and initial conditions.
        Simulate the system response at time points `t` due to loads
        `F` and with initial displacements `d0` and velocities `v0`.
        Arguments
        ---------
        t : 1darray
            Time points to evaluate the system response.
        F : 2darray, optional
            Load matrix where each column corresponds to time points in
            `t` and each row is the load applied to a system dof. Fij is
            then the load applied to dof `i` at time `j`. Zeros is assumed
            if None.
        d0, v0 : 1darray, optional
            Initial displacment and velocity vector. Zeros is assumed
            if None.
        Returns
        -------
        A, V, D : 2darray
            Acceleration, velocity and displacement vector for the system.
        """
        n = self.n
        d0 = np.zeros(n) if d0 is None else d0
        v0 = np.zeros(n) if v0 is None else v0
        x0 = np.r_[v0, d0]

        sys = signal.StateSpace(*self.get_state_space_matrices())
        U = np.zeros((t.size, n)) if F is None else F.T

        _, y, _ = signal.lsim(sys, U, t, X0=x0)
        y = y.T
        A = y[:n, :]
        V = y[n:2*n, :]
        D = y[2*n:, :]
        return A, V, D


def rmfd2ss(N, D):
    """Convert RMFD model to state space model
    Convert a right matrix fraction description (RMFD)
    matrices N and D to state space matrices A, B, C, D.
    See [Reynders2012].
    Arguments
    ---------
    N, D : 3darrays
        Numerator and denominator matrices for RMFD model.
    Returns
    -------
    A, B, C, D : 2darrays
        State space matrices.
    References
    ----------
    [Reynders2012] Reynders, E., 2012. System Identification Methods
        for (Operational) Modal Analysis: Review and Comparison. Arch
        Computat Methods Eng 19, 51–124.
        https://doi.org/10.1007/s11831-012-9069-x
    """
    n, l, m = N.shape
    A = np.zeros((n*m, n*m))
    A[m:, :-m] = np.eye((n-1)*m)
    C = np.zeros((l, n*m))
    Np = N[-1]
    Dp = D[-1]
    for k, (Dk, Nk) in enumerate(zip(D[:-1][::-1], N[:-1][::-1])):
        DpDk = np.linalg.solve(Dp, Dk)
        A[:m, k*m:(k+1)*m] = -DpDk
        C[:, k*m:(k+1)*m] = Nk - Np@DpDk
    B = np.zeros((n*m, m))
    B[:m, :] = np.linalg.solve(Dp, np.eye(m))
    D = np.linalg.solve(Dp.T, Np.T).T
    return A, B, C, D


class Mode(object):
    def __init__(self, eigenvalue, eigenvector):
        """Mode converts eigenvalue/vector to vibration mode characteristics
        A mode defines a single degree of freedom dynamic system
        with frequency (f and w) and damping (xi) characteristics and mode
        shape vector (v).
        Arguments
        ---------
        eigenvalue : complex
            Eigenvalue (in continous time) of the mode.
        eigenvector : 1darray[float or complex]
            Eigenvector or modal vector.
        """
        self.eigenvalue = eigenvalue
        self.eigenvector = eigenvector

    @property
    def v(self):
        return self.eigenvector / np.linalg.norm(self.eigenvector, 2)

    @property
    def w(self):
        return self.eigenvalue.imag / np.sqrt(1-self.xi**2)

    @property
    def wd(self):
        return self.eigenvalue.imag

    @property
    def xi(self):
        u = self.eigenvalue
        return -u.real / np.abs(u)

    @property
    def f(self):
        return self.w / (2.0*np.pi)

    @property
    def fd(self):
        return self.wd / (2.*np.pi)

    @property
    def mean_phase_colinearity(self):
        return modal_phase_collinearity(self.v)

    @property
    def mpc(self):
        return self.mean_phase_colinearity

    @property
    def mean_phase(self):
        return mean_phase(self.v)

    @property
    def mp(self):
        return self.mean_phase

    @property
    def mean_phase_deviation(self):
        return mean_phase_deviation(self.v)

    @property
    def mpd(self):
        return self.mean_phase_deviation

    @classmethod
    def find_modes_from_ss(cls, A, C, fs):
        """Return modes from the (discrete) system matrices A and C
        This method finds all modes from the discrete state space system
        matrices A and C.
        Arguments
        ---------
        A : 2darray
            Discrete time state space matrix
        C : 2darray
            Output influence matrix
        fs : float
            Sampling rate
        Returns
        -------
        list
            List of modes (Mode objects)
        """
        lr, Q = np.linalg.eig(A)
        u = fs*np.log(lr)
        Phi = C.dot(Q)
        return [cls(ui, q) for ui, q in zip(u, Phi.T)]

    @classmethod
    def find_modes_from_rmfd(cls, N, D, fs):
        """Return modes from a the right matrix fraction description (RMFD).
        This method returns modes from the RMFD numerator and denominator
        matrices N and D.
        Arguments
        ---------
        N, D : 3darrays
            Numerator and denominator matrices for RMFD.
        fs : float
            Sampling rate
        Returns
        -------
        list
            List of modes (Mode objects)
        """
        A, B, C, D = rmfd2ss(N, D)
        return cls.find_modes_from_ss(A, C, fs)


qos-pypackage/csi-beam/sys-id.py
"""Console script for qos."""
#%matplotlib nbagg


import matplotlib.pyplot as plt
import numpy as np
import sysid
import scipy.signal
#----------------------------------------

data19 = np.load("node19data.npz")
y19 = data19["y"]
fs19 = data19["fs"]
true_f19 = data19["true_frequencies"]
true_xi19 = data19["true_damping"]
true_modeshapes19 = data19["true_modeshapes"]
sensor_pos19=data19["sensor_pos"]

s=np.array([0,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1])
data = np.load("node19-9data.npz")
y = data["y"]
fs = data["fs"]
true_f = data["true_frequencies"]
true_xi = data["true_damping"]
true_modeshapes = data["true_modeshapes"]
sensor_pos=data["sensor_pos"]
ssid = sysid.CovarianceDrivenStochasticSID(y, fs)
"""
_,T=(np.shape(y19))/fs
t = np.arange(0., T, 1 / fs)
plt.figure("Displacement measured node 1 and 10")
plt.plot(t, y19[9], label="node 10")
plt.plot(t, y19[0], label="node 1")
plt.legend()
plt.xlabel("Time (s)")
plt.ylabel("Amplitude")

plt.figure("PSD of Displacement at 1 and 10 node")
for yi in [y19[0], y19[9]]:
    freqs, Pyy = scipy.signal.csd(yi,yi, fs, nperseg=2 ** 12)
    plt.semilogy(freqs, Pyy)
    plt.xlim(0,50)
plt.legend(['node 1','mode 10'])
plt.ylabel('PSD')
plt.xlabel('Frequency (Hz)')
plt.show()"""

modes = dict()
for i, order in enumerate(range(30, 100, 2)):
    A, C, G, R0 = ssid.perform(order, 20)
    modes[order] = sysid.Mode.find_modes_from_ss(A, C, fs)

sd = sysid.StabilizationDiagram()
sd.plot(modes)
f, psd = ssid.psdy(nperseg=2**12)
#sd.axes_psd.semilogy(f, np.trace(np.abs(psd)), color=(0., 0., 0., .5), lw=0.3)
for i in range(2):
    freqs, Pyy = scipy.signal.csd(y[i],y[i], fs, nperseg=2 ** 12)
    sd.axes_psd.semilogy(freqs, Pyy,color=(0., 0., 0.+i, .5), lw=0.3)

plt.show()

modes = sd.picked_modes

fig = plt.figure("Damping estimate")
axd = fig.add_axes((0.1, 0.1, .8, .8))
axd.set(xlabel='Frequency', ylabel='Damping ratio', title='Estimated and true frequency and damping',
        ylim=(0, .10)
        )
figmodes, axes = plt.subplots(ncols=5, nrows=4, dpi=144)
for n in range(true_f19.size):
    ax = axes.flatten()[n]
    un19 = true_modeshapes19[n]
    unn = true_modeshapes19[n, :true_f.size]
    fn19 = true_f19[n]
    xin19 = true_xi19[n]
    nmax = np.argmax([sysid.utils.modal_assurance_criterion(mode.v, unn) for mode in modes])
    mode = modes[nmax]
    line, = axd.plot(fn19, xin19, '.')
    line1, = axd.plot(mode.f, mode.xi, 'x',mec=line.get_color(), mfc=(0, 0, 0, 0))

    v_true = np.r_[0., un19, 0.]
    ax.plot(sensor_pos19, v_true, label='True', lw=0.1, marker='.', markersize=1)
    ax.tick_params(axis='x', labelsize=5)
    ax.tick_params(axis='y', labelsize=5)
    ax.axhline(0., color=(0, 0, 0, .3))

    ax.set_title(f"Mode {n + 1}", x=0.5, y=0.8, fontsize=6, fontweight="bold")
    ax.axis('on')
    ax.set_xlim(0, 1200)
    ax.set_ylim(-1, 1)
    plt.grid(True)
"""
res=[]
for n in range(true_f.size):
    ax = axes.flatten()[n]
    un = true_modeshapes[n]
    unn = true_modeshapes19[n]*s[1:-1]
    unn = unn[np.nonzero(unn)]
    nmax = np.argmax([sysid.utils.modal_assurance_criterion(mode.v, unn) for mode in modes])
    mode = modes[nmax]
    vn = np.r_[0., unn, 0.]
    v = np.r_[0., mode.v, 0.]
    v = sysid.utils.modal_scale_factor(v, vn) * v
    ax.plot(sensor_pos, v.real, label='Estimated', lw=0.1, marker='x', markersize=1)
    if n == 2:
        ax.legend(bbox_to_anchor=(.5, 1.20), loc='lower center', ncol=1)
        axd.legend(['True','Estimated'] )

    mac = sysid.utils.modal_assurance_criterion(unn, mode.v)
    msf=sysid.utils.modal_scale_factor(unn, mode.v)
    res.append([n, mac * 100, msf, ])

"""
#print(res)
plt.show()
#np.savetxt('mac19-9.txt',res,delimiter=',')
print("done")
